### ISSUE TITLE
[Feature Request] ass.h - Add method ass_get_fonts

### ISSUE THREAD

**Issue by moi15moi** [libass/libass#642]
### Feature Request
Would it be conceivable to add the method: ``ass_get_fonts(ASS_Track *track)``?

### What would it do
This method would return all the font path that libass would use to render one .ass file.
In brief, it would be an FontCollector like Aegisub, but directly in libass.

**Comment by astiob** [comment]
My thoughts (not necessarily a rejection; discussion welcome):

libass uses fonts from a variety of sources, most prominently fonts already embedded in the video file and in the subtitle file, which have no paths.

And libass’s goal when it searches for fonts is very much not what a font collector’s goal should be. libass’s choice can be suboptimal (but during playback, any choice is better than nothing), and even if you were to grab all of the currently selected fonts, put them in an MKV or ASS file and play it back using the same libass on a different computer, you might see some of the attached fonts ignored and different fonts from the system picked instead.

For a good subtitle release, you ideally want the subtitles to never use font fallback and to always select fonts exactly, with no room for ambiguity. In my opinion, a font collector should strive to verify these properties and complain to the user if their subtitles are ambiguous. libass currently has no way of doing this, although something like an “authoring mode” has been proposed in the past and might make this viable.

**Comment by moi15moi** [comment]
**Reason why I think an FontCollector would need to be part of libass.**

Libass has a algorithm to make a choice when it pick font.
Any FontCollector need to mimic this algorith to know which font libass need.

Even if the algorithm is not supposed to change, it can (see #613).
So, a FontCollector could be obsolete. If libass would propose a FontCollector, it would always be up to date.

**Currently, how libass pick font?**
```
1. ass_render_frame (ass_render)
2. ass_render_event (ass_render)
3. ass_shaper_find_runs (ass_shaper)
4. ass_font_get_index (ass_font)
5. add_face (ass_font)
6. ass_font_select (ass_fontselect)
```

We could add ``fontpath`` to the struct ``ASS_FontDesc``. 
The method ``add_face`` would set the fontpath.

**Implementation problematic**
- What do we do with font added with ass_add_font (font embedded in the video)? Do we ignore them?
- How do we ignore the "default" font?
- How to reuse the code from ``ass_render_frame`` to ``ass_shaper_find_runs``? Some part of the code are needed.

> although something like an “authoring mode” has been proposed...

Is that discussion is on github? I would like to get more information about this possibility.

**Comment by astiob** [comment]
> Even if the algorithm is not supposed to change, it can (see #613).
> So, a FontCollector could be obsolete. If libass would propose a FontCollector, it would always be up to date.

In my eyes, this is precisely _more reason_ to avoid this in libass. A font collector’s task is to gather fonts that will be sufficient and necessary for playback in stable/future renderers, not in a particular bugged version of libass.

And as mentioned, libass makes no attempt to produce a stable set of fonts as output that can be fed back in. This would have to be new functionality, not just a new API.

> We could add `fontpath` to the struct `ASS_FontDesc`.

libass’s internal fonts already store paths… if they are known/used. And for that matter, there’s no guarantee that they are: on some platforms, paths are unknown and fonts are streamed in-memory from the system’s font provider.

> > although something like an “authoring mode” has been proposed...
> 
> Is that discussion is on github? I would like to get more information about this possibility.

#78

**Comment by astiob** [comment]
> A font collector’s task is to gather fonts that will be sufficient and necessary for playback in stable/future renderers, not in a particular bugged version of libass.

Ideally, of course, the font collector would detect any ambiguity between current libass and VSFilter/future libass and either produce a font set that somehow handles both or ask the user to adjust the fonts to remove the ambiguity. After all, there are two potentially conflicting goals:

* Persist the way the file looks on the current machine, right now, with the current renderer.
* Make sure the file looks the same way in all renderers now and in the future.

**Comment by moi15moi** [comment]
>> Even if the algorithm is not supposed to change, it can (see https://github.com/libass/libass/issues/613).
> So, a FontCollector could be obsolete. If libass would propose a FontCollector, it would always be up to date.
> 
> In my eyes, this is precisely more reason to avoid this in libass. A font collector’s task is to gather fonts that will be sufficient and necessary for playback in stable/future renderers, not in a particular bugged version of libass.

Any other tool than libass will mimic libass algorithm or it will use VSFilter (I don't know how Aegisub collect the font).
Libass is the only one that really know how he picks up the font, so in my opinion, he should be the one to collect the font.

>>> although something like an “authoring mode” has been proposed...
>>>
>>Is that discussion is on github? I would like to get more information about this possibility.
>>
>https://github.com/libass/libass/issues/78

It would make more sense to be able to collect the font in this famous "strict mode".

**Comment by TheOneric** [comment]
I agree with astiob. Emulating and/or relying of libass or any other renderers font *fallback* logic does not strike me as a good idea for a authoring font collector. Even for the same libass commit, just bundling all fallback fonts won't ensure the rendering is identical, as fallbacks are more lenient then memory font matches and are retrieved from the system font provider which may have synthesised or normalised some properties, which ofc won't happen or happen differently with libass own memory font logic (see e.g. #634 and #509).

Imho, an ideal font collector should only accepted exact name and property matches and otherwise error out with a helpful message informing users what fonts don't have exact matches and perhaps suggest possible fixes. On top of just ensuring exat matches, there's a bit of additional checks required to ensure it'll work in all renderers. Some of those are implemented in libass some aren't but in either case libass won't generate error messages helpful for font collector users. I don't expect this to change with an “authoring mode”, afterall libass is primarily still an renderer not a general authoring tool. So in the “authoring mode” future, it might be possible to collect all required fonts when the file is already fully correct and portable, but if it isn't it won't make a pleasant to use font collector. Those checks are relatively stable though and not too complicated if I didn't forget about something, so it shouldn't be too hard to implement in a font collector.
 - VSFilter can only deal with font names up to 31 characters. Reject fonts with longer names.
 - VSFilter can afaik only deal with TrueType and OpenType fonts, so reject anything else (e.g. Postscript, Graphite, ...)
 - VSFilter running on not up-to-date Windows cannot deal with Variable Fonts; for libass it is inconsistent and depends on the font provider.
   Currently: reject variable fonts.
   In the future when libass can deal with them like VSFilter: warn users there might be issues with older libass or Windows versions without error.
 - VSFilter running on not up-to-date Windows cannot deal with any non-bitmap colour emojis, neither can current libass.
   Currently: reject non-bitmap colour emoji.
   In the future: warn about old Windows and libass versions without error

**Comment by astiob** [comment]
> VSFilter running on not up-to-date Windows cannot deal with any non-bitmap colour emojis

Colour emoji? Even on up-to-date Windows, what can VSFilter possibly do with colour emoji? It's still a monochromatic renderer.

**Comment by TheOneric** [comment]
Oops, mixed things up regarding colour emojis; it appears no renderer currently supports colour fonts. Now corrected, thanks!
